File: src/main/java/no/nav/foreldrepenger/stønadskonto/regelmodell/grunnlag/BeregnKontoerGrunnlag.java
package no.nav.foreldrepenger.stønadskonto.regelmodell.grunnlag;

import java.time.LocalDate;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

import no.nav.foreldrepenger.stønadskonto.regelmodell.StønadskontoKontotype;
import no.nav.fpsak.nare.doc.RuleDocumentationGrunnlag;

@RuleDocumentationGrunnlag
public class BeregnKontoerGrunnlag {

    /*
     * Normalt velges konti ut fra fødselsdato/termindato/omsorgsdato.
     * Normalt vil endringer tre i kraft "for nye tilfelle" - som enten går på familiehendelsedato eller på første uttaksdato.
     * For regler som gjelder begge foreldrene brukes gjerne familiehendelsesdato. Andre regler som er mer individuelt orientert ser gjerne på uttaksdato
     *
     * For tilfelle terminbasert søknad så vil man på søknadstidspunktet ikke nødvendigvis vite om fødselsdato tilsier annet regelverk enn termin.
     * Dessuten trer nye regler først i kraft når dagens dato har passert ikraftredelsesdato for endrete regler - må innvilges på gamle regler inntil nye trer i kraft
     *
     * Parameter regelvalgsdato settes kun når man ønsker å "overstyre" familiehendelsedato for regelvalg og kan brukes i utviklingsmiljø + produksjon fram til ikrafttredelse.
     */
    private final Map<StønadskontoKontotype, Integer> tidligereUtregning = new LinkedHashMap<>();

    private LocalDate regelvalgsdato;
    private Dekningsgrad dekningsgrad;
    private Rettighetstype rettighetstype;
    private Brukerrolle brukerrolle;

    // For utregning av tilleggsdager
    private int antallBarn = 1;
    private LocalDate fødselsdato;
    private LocalDate termindato;
    private LocalDate omsorgsovertakelseDato;

    // For utregning av tilkommet minsterett
    private boolean morHarUføretrygd;
    private LocalDate familieHendelseDatoNesteSak;

    private BeregnKontoerGrunnlag() {
    }

    public int getAntallBarn() {
        return antallBarn;
    }

    public boolean isMorRett() {
        return isBeggeRett() || Brukerrolle.MOR.equals(brukerrolle);
    }

    public boolean isFarRett() {
        return isBeggeRett() || !Brukerrolle.MOR.equals(brukerrolle);
    }

    public boolean isBeggeRett() {
        return Rettighetstype.BEGGE_RETT.equals(rettighetstype);
    }

    public Dekningsgrad getDekningsgrad() {
        return dekningsgrad;
    }

    public Map<StønadskontoKontotype, Integer> getTidligereUtregning() {
        return tidligereUtregning;
    }

    public Rettighetstype getRettighetstype() {
        return rettighetstype;
    }

    public Brukerrolle getBrukerrolle() {
        return brukerrolle;
    }

    public boolean isMorHarUføretrygd() {
        return morHarUføretrygd;
    }

    public LocalDate getFamilieHendelseDatoNesteSak() {
        return familieHendelseDatoNesteSak;
    }

    public boolean isBareFarHarRett() {
        return Rettighetstype.BARE_SØKER_RETT.equals(rettighetstype) && !Brukerrolle.MOR.equals(brukerrolle);
    }

    public boolean isAleneomsorg() {
        return Rettighetstype.ALENEOMSORG.equals(rettighetstype);
    }

    public boolean isGjelderFødsel() {
        return erFødsel();
    }

    public LocalDate getFamilieHendelseDato() {
        return Optional.ofNullable(omsorgsovertakelseDato)
            .or(this::getFødselsdato)
            .or(this::getTermindato)
            .orElse(null);
    }

    private Optional<LocalDate> getRegelvalgsdato() {
        return Optional.ofNullable(regelvalgsdato);
    }

    public LocalDate getKonfigurasjonsvalgdato() {
        return getRegelvalgsdato().orElseGet(this::getFamilieHendelseDato);
    }


    public boolean erFødsel() {
        return fødselsdato != null || termindato != null;
    }

    public Optional<LocalDate> getFødselsdato() {
        return Optional.ofNullable(fødselsdato);
    }

    public Optional<LocalDate> getTermindato() {
        return Optional.ofNullable(termindato);
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private final BeregnKontoerGrunnlag kladd = new BeregnKontoerGrunnlag();

        public Builder regelvalgsdato(LocalDate regelvalgsdato) {
            kladd.regelvalgsdato = regelvalgsdato;
            return this;
        }

        public Builder tidligereUtregning(Map<StønadskontoKontotype, Integer> tidligereUtregnet) {
            kladd.tidligereUtregning.putAll(tidligereUtregnet);
            return this;
        }

        public Builder antallBarn(int antallBarn) {
            kladd.antallBarn = antallBarn;
            return this;
        }

        public Builder rettighetType(Rettighetstype rettighetstype) {
            kladd.rettighetstype = rettighetstype;
            return this;
        }

        public Builder brukerRolle(Brukerrolle brukerrolle) {
            kladd.brukerrolle = brukerrolle;
            return this;
        }

        public Builder dekningsgrad(Dekningsgrad dekningsgrad) {
            kladd.dekningsgrad = dekningsgrad;
            return this;
        }

        public Builder fødselsdato(LocalDate dato) {
            kladd.fødselsdato = dato;
            return this;
        }

        public Builder termindato(LocalDate dato) {
            kladd.termindato = dato;
            return this;
        }

        public Builder omsorgsovertakelseDato(LocalDate dato) {
            kladd.omsorgsovertakelseDato = dato;
            return this;
        }

        public Builder morHarUføretrygd(boolean morHarUføretrygd) {
            kladd.morHarUføretrygd =  morHarUføretrygd;
            return this;
        }

        public Builder familieHendelseDatoNesteSak(LocalDate familieHendelseDatoNesteSak) {
            kladd.familieHendelseDatoNesteSak = familieHendelseDatoNesteSak;
            return this;
        }

        public BeregnKontoerGrunnlag build() {
            Objects.requireNonNull(kladd.rettighetstype, "rettighetType");
            Objects.requireNonNull(kladd.brukerrolle, "brukerRolle");
            if (kladd.fødselsdato == null && kladd.termindato == null && kladd.omsorgsovertakelseDato == null) {
                throw new IllegalArgumentException("Forventer minst en familiehendelsedato");
            }
            return kladd;
        }
    }
}

File: stonadskonto/src/grunnlag/BeregnKontoerGrunnlag.ts
import { Dekningsgrad } from './Dekningsgrad';
import { Rettighetstype } from './Rettighetstype';
import { Brukerrolle } from './Brukerrolle';

export class BeregnKontoerGrunnlag {
  public regelvalgsdato: Date | null = null;
  public dekningsgrad: Dekningsgrad | null = null;
  public rettighetstype: Rettighetstype | null = null;
  public brukerrolle: Brukerrolle | null = null;
  public antallBarn: number = 1;
  public fødselsdato: Date | null = null;
  public termindato: Date | null = null;
  public omsorgsovertakelseDato: Date | null = null;
  public morHarUføretrygd: boolean = false;
  public familieHendelseDatoNesteSak: Date | null = null;

  constructor() {}

  getAntallBarn(): number {
    return this.antallBarn;
  }

  isMorRett(): boolean {
    return this.isBeggeRett() || this.brukerrolle === Brukerrolle.MOR;
  }

  isFarRett(): boolean {
    return this.isBeggeRett() || this.brukerrolle !== Brukerrolle.MOR;
  }

  isBeggeRett(): boolean {
    return this.rettighetstype === Rettighetstype.BEGGE_RETT;
  }

  getDekningsgrad(): Dekningsgrad | null {
    return this.dekningsgrad;
  }

  getRettighetstype(): Rettighetstype | null {
    return this.rettighetstype;
  }

  getBrukerrolle(): Brukerrolle | null {
    return this.brukerrolle;
  }

  isMorHarUføretrygd(): boolean {
    return this.morHarUføretrygd;
  }

  getFamilieHendelseDatoNesteSak(): Date | null {
    return this.familieHendelseDatoNesteSak;
  }

  isBareFarHarRett(): boolean {
    return this.rettighetstype === Rettighetstype.BARE_SØKER_RETT && this.brukerrolle !== Brukerrolle.MOR;
  }

  isAleneomsorg(): boolean {
    return this.rettighetstype === Rettighetstype.ALENEOMSORG;
  }

  isGjelderFødsel(): boolean {
    return this.erFødsel();
  }

  getFamilieHendelseDato(): Date | null {
    return this.omsorgsovertakelseDato || this.fødselsdato || this.termindato || null;
  }

  getKonfigurasjonsvalgdato(): Date | null {
    return this.regelvalgsdato || this.getFamilieHendelseDato();
  }

  erFødsel(): boolean {
    return this.fødselsdato !== null || this.termindato !== null;
  }

  getFødselsdato(): Date | null {
    return this.fødselsdato;
  }

  getTermindato(): Date | null {
    return this.termindato;
  }

  static builder(): Builder {
    return new Builder();
  }
}

export class Builder {
  private kladd = new BeregnKontoerGrunnlag();

  regelvalgsdato(regelvalgsdato: Date | null): Builder {
    this.kladd.regelvalgsdato = regelvalgsdato;
    return this;
  }

  antallBarn(antallBarn: number): Builder {
    this.kladd.antallBarn = antallBarn;
    return this;
  }

  rettighetType(rettighetstype: Rettighetstype): Builder {
    this.kladd.rettighetstype = rettighetstype;
    return this;
  }

  brukerRolle(brukerrolle: Brukerrolle): Builder {
    this.kladd.brukerrolle = brukerrolle;
    return this;
  }

  dekningsgrad(dekningsgrad: Dekningsgrad): Builder {
    this.kladd.dekningsgrad = dekningsgrad;
    return this;
  }

  fødselsdato(dato: Date | null): Builder {
    this.kladd.fødselsdato = dato;
    return this;
  }

  termindato(dato: Date | null): Builder {
    this.kladd.termindato = dato;
    return this;
  }

  omsorgsovertakelseDato(dato: Date | null): Builder {
    this.kladd.omsorgsovertakelseDato = dato;
    return this;
  }

  morHarUføretrygd(morHarUføretrygd: boolean): Builder {
    this.kladd.morHarUføretrygd = morHarUføretrygd;
    return this;
  }

  familieHendelseDatoNesteSak(familieHendelseDatoNesteSak: Date | null): Builder {
    this.kladd.familieHendelseDatoNesteSak = familieHendelseDatoNesteSak;
    return this;
  }

  build(): BeregnKontoerGrunnlag {
    if (!this.kladd.rettighetstype) {
      throw new Error('rettighetType is required');
    }
    if (!this.kladd.brukerrolle) {
      throw new Error('brukerRolle is required');
    }
    if (!this.kladd.fødselsdato && !this.kladd.termindato && !this.kladd.omsorgsovertakelseDato) {
      throw new Error('At least one family event date is required (fødselsdato, termindato, or omsorgsovertakelseDato)');
    }
    return this.kladd;
  }
}

